[YDS-#216] Component - TabBar 구현

Gael-Android opened the pull request:
## Summary
- Tab, FixedTabBar, ScrollableTabBar 구현
- 머티리얼 의존성 제거
- 커스텀 레이아웃 구현

## Describe your changes

### FixedTabBar
![FixedTabBar ani](https://github.com/yourssu/YDS-Android/assets/84790707/8ca76c2c-6f9a-4d5d-bb08-6fbe9bfabc49)

### ScrollableTabBar
![ScrollTabBar ani](https://github.com/yourssu/YDS-Android/assets/84790707/cd23aaa7-d77d-4199-8713-3523f5c764c4)



## Issue
- #216

Gael-Android committed:
TabBar,Tab 구현중

Gael-Android committed:
ScrollableTabBar 구현중

Gael-Android committed:
FixedTabBar, ScrollableTabBar, Tab 구현완료

Gael-Android committed:
ScrollableTabBar Update

Gael-Android committed:
Revert "ScrollableTabBar Update"

This reverts commit 534a429b674bbac3b0e27050079267dadd5e52b4.

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Tab.kt:
```diff
@@ -0,0 +1,143 @@
+            verticalArrangement = Arrangement.Center,
+            horizontalAlignment = Alignment.CenterHorizontally
+        ) {
+            TabLayout(
+                text = text,
+            )
+        }
```
혹시 Box의 contentAlignment 안 쓰시고 이런 방식으로 직접 계산해서 넣어주신 이유가 있으실까요? 대충 코드만 봐서는 잘 모르겠어서요!

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/TabBar.kt:
```diff
@@ -0,0 +1,576 @@
+                Modifier.tabIndicatorOffset(tabPositions[selectedTabIndex])
+            )
+        },
+    divider: @Composable @UiComposable () -> Unit = @Composable {
+        TabBarDefaults.Divider()
+    },
```
YDS에서 divider는 고정이라서 따로 외부에서 받아줄 필요는 없을 것 같아요!

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/TabBar.kt:
```diff
@@ -0,0 +1,576 @@
+                    slotId = TabSlots.Tabs,
+                    content = tabs
+                )
+                val tabCount = measurableTabs.size
```
YDS 명세에 최소 2개, 최대 5 탭만 사용해달라는 조건이 있어서 require 같은걸로 체크해주는 코드 넣어주시면 감사하겠습니다

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/TabBar.kt:
```diff
@@ -0,0 +1,576 @@
+                .clipToBounds()
+        ) { constraints ->
+            val minTabWidth = ScrollableTabWidth.roundToPx()
+            val edgeRadding = edgePadding.roundToPx()
```
엇 edgeRadding 이거 오타인가요?

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/TabBar.kt:
```diff
@@ -0,0 +1,576 @@
+    }
+}
+
+private val ScrollableTabWidth = 88.dp
```
상수는 Defaults 클래스나, 파일 하단에 모아두면 보기 편할 것 같아요

cometj03 reviewed (commented):
수고하셨습니다! 머티리얼 코드 참고하셨다고 해도 상당한데요

근데 전체적으로 구현하신 걸 보면 탭이랑 인디케이터랑 따로 그리는 것 같더라구요. 혹시 추후에 인디케이터 애니메이션도 염두에 둔 구현인걸까요? YDS에 나와있는 건 그냥 선택되어있으면 인디케이터가 보이고, 그렇지 않으면 안 보이는 단순한 방식이어서 따로 그릴 필요까진 없을 것 같다는 생각이 들어서요!

우선은 코멘트 드린 부분만 살짝 수정해주시면 될 것 같습니다 👍

Gael-Android committed:
리뷰 반영
- Tab 의 커스텀 Layout 제거 -> Box
- TabBar 들의 매개변수로 들어오는 Divider, Indicator 제거 -> 커스텀 X
- FixedTabBar 의 탭 수가 최소 2개, 최대 5개라 require 로 확인
- 오타 수정 (edgeRadding -> edgePadding)
- 상수들 파일 하단, Default 클래스 내부로 이동

Gael-Android commented on compose/src/main/java/com/yourssu/design/system/compose/component/Tab.kt:
```diff
@@ -0,0 +1,116 @@
+    CompositionLocalProvider(
+        LocalContentColor provides color.copy(alpha = 1f),
+    ) {
+        Column(
+            modifier = modifier
+                .selectable(
+                    selected = selected,
+                    interactionSource = interactionSource,
+                    indication = null,
+                    enabled = true,
+                    onClick = onClick
+                ),
+            verticalArrangement = Arrangement.Center,
+            horizontalAlignment = Alignment.CenterHorizontally
+        ) {
```
그러네요

cometj03 commented:
아 그리고 version.properties 파일에서 하위 버전 하나 올려주세요! (슬랙 안드 채널 참고)

Gael-Android committed:
Tab 불필요 Column 제거, version 올리기

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Tab.kt:
```diff
@@ -0,0 +1,109 @@
+    selected: Boolean,
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier,
+    text: @Composable (() -> Unit),
```
외부에서 텍스트 컴포저블을 넣어줄 때 yds tab 명세에 따라서 스타일 지정해주는 게 번거로울 것 같아요. 그냥 단순히 String 하나 받는 건 어떨까요?

cometj03 commented on compose/src/main/java/com/yourssu/design/system/compose/component/Tab.kt:
```diff
@@ -0,0 +1,109 @@
+) {
+    val color = if (selected) selectedContentColor else unselectedContentColor
+
+    CompositionLocalProvider(
+        LocalContentColor provides color.copy(alpha = 1f),
+    ) {
```
위에 text 함수 받는 게 사라진다면 이 코드도 없어도 될 것 같네요!

cometj03 reviewed (commented):
리뷰 한 번에 해야 하는데 띄엄띄엄 드려서 죄송합니다ㅠ 의견만 간단히 남겨주시고 나머지는 이따 회의에서 의논해보면 좋을 것 같아요!

Gael-Android committed:
Tab 텍스트 입력을 컴포저블이 아닌 텍스트로 입력

Gael-Android commented:
컴포넌트 개발은 5~7일정도 소요되네요
